x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/TransformIndexer.java

```diff
@@ -468,25 +468,32 @@ protected void onAbort() {
 
     synchronized void handleFailure(Exception e) {
         logger.warn(new ParameterizedMessage("[{}] transform encountered an exception: ", getJobId()), e);
-        if (handleCircuitBreakingException(e)) {
-            return;
-        }
-
-        if (isIrrecoverableFailure(e) || context.getAndIncrementFailureCount() > context.getNumFailureRetries()) {
-            String failureMessage = isIrrecoverableFailure(e)
-                ? "task encountered irrecoverable failure: " + e.getMessage()
-                : "task encountered more than " + context.getNumFailureRetries() + " failures; latest failure: " + e.getMessage();
-            failIndexer(failureMessage);
+        ElasticsearchException unwrappedException = SearchExceptionRootCauseFinder.getRootCauseElasticsearchException(e);
+
+        if (unwrappedException instanceof CircuitBreakingException) {
+            handleCircuitBreakingException((CircuitBreakingException) unwrappedException);
+        } else if (unwrappedException instanceof ScriptException) {
+            handleScriptException((ScriptException) unwrappedException);
+            // irrecoverable error without special handling
+        } else if (unwrappedException instanceof IndexNotFoundException
+            || unwrappedException instanceof AggregationResultUtils.AggregationExtractionException
+            || unwrappedException instanceof TransformConfigReloadingException) {
+            failIndexer("task encountered irrecoverable failure: " + e.getMessage());
+        } else if (context.getAndIncrementFailureCount() > context.getNumFailureRetries()) {
+            String failure = unwrappedException != null ? unwrappedException.getDetailedMessage() : e.getMessage();
```

Comment:
I think it would be beneficial to add a `getDetailedMessage` method to the transforms ExceptionsHelper class. It could do a `if(exception instanceof ElasticsearchException)` and then cast and return the detailed message. And `exception.getMessage` otherwise.